core/meta/src/TClass.cxx

```diff
@@ -4359,29 +4359,33 @@ Int_t TClass::GetNmethods()
 
 TVirtualStreamerInfo* TClass::GetStreamerInfo(Int_t version /* = 0 */) const
 {
-   TVirtualStreamerInfo *guess = fLastReadInfo;
-   if (guess && guess->GetClassVersion() == version) {
-      // If the StreamerInfo is assigned to the fLastReadInfo, we are
-      // guaranteed it was built and compiled.
-      return guess;
-   }
+   TVirtualStreamerInfo *sinfo = fLastReadInfo;
 
-   R__LOCKGUARD(gInterpreterMutex);
+   // Version 0 is special, it means the currently loaded version.
+   // We need to set it at the beginning to be able to guess it correctly.
 
-   // Handle special version, 0 means currently loaded version.
-   // Warning:  This may be -1 for an emulated class.
-   // If version == -2, the user is requested the emulated streamerInfo
-   // for an abstract base class even though we have a dictionary for it.
-   if (version == 0) {
+   if (version == 0)
       version = fClassVersion;
-   }
-   Int_t ninfos = fStreamerInfo->GetSize();
-   if ((version < -1) || (version >= ninfos)) {
+
+   // If the StreamerInfo is assigned to the fLastReadInfo, we are
+   // guaranteed it was built and compiled.
+   if (sinfo && sinfo->GetClassVersion() == version)
```

Comment:
Reminder of the discussion on this line, in the previous PR

Reply:
>  	
> @Axel-Naumann
> Axel-Naumann on Jun 29 Owner
> 
> That's a change in behavior (version might have been set to fClassVersion after your change but not before); could you document it in the commit log if it was intentional?
> @amadio
> amadio on Jun 29 • edited Member
> 
> This is intentional. When the requested version is 0, it really means the current version (which supposedly isn't itself equal to 0), so this let's the detection work properly and return the version already available (that's the point of this check). However, before the check would always fail, because sinfo->GetClassVersion() never returns 0 (at least that's how I interpreted it, correct me if I'm wrong).
> @Axel-Naumann
> Axel-Naumann on Jun 29 Owner
> 
>     (which supposedly isn't itself equal to 0)
>     ...
>     (at least that's how I interpreted it, correct me if I'm wrong)
> 
> These are assumptions; they suggest that we are probably better off waiting for Philippe with these changes?
> @amadio
> amadio on Jun 29 Member
> 
> Yes, definitely, I will not merge this before Philippe is back. I just want to give others a chance to review immediately to address any problems found before he comes back.
> @pcanal
> pcanal on Jul 3 Owner
> 
> This looks fine but really hinge on whether the line number 4366 (in new file) is thread safe or not.
> @amadio
> amadio on Jul 4 Member
> 
> That line is apparently not thread-safe, due to false sharing (lack of alignment of fClassVersion), but real problems depend on if there is any pattern where modifications to the class version can actually happen when this function is called. As commented above, we may need to go back to taking the lock despite the performance penalty. On the other hand, we need a better long term solution for our thread-safety issues. On one side we need a test suite to catch thread-safety problems like these, and on the other, we need a better way than taking locks and using atomics.
>

Reply:
Note that fClassVersion is not (yet?) an atomic.  Note the implementation of TClass::GetClassVersion
Note the existence of TClass::SetClassVersion which is only use in ROOT::ResetClassVersion which relies on the atomic fVersionUsed to allow or disallow the change.   This feature is usually triggered by the use of the macro RootClassVersion (so the SetClassVersion is usually executed at library load time).

And a cursory look a the code mentions seems to indicates that the existing situation is already not really thread-safe ... (missing lock taking in the above mention function, missing setting of fVersionUsed in direct use of fClassVersion in TClass.cxx and/or fClassVersion not being atomic ....)

Reply:
Indeed, we need a better solution for this kind of problem, and I suggested using [RCU](http://www.rdrop.com/~paulmck/RCU) before, which should be much better in terms of performance than using locks.